src/midiEditorCommands.cpp

```diff
@@ -304,49 +407,46 @@ const string getMidiNoteName(MediaItem_Take *take, int pitch, int channel) {
 
 // Returns the indexes of the first and last notes in a chord in a given direction.
 pair<int, int> findChord(MediaItem_Take* take, int direction) {
-	int count;
-	MIDI_CountEvts(take, &count, NULL, NULL);
 	double now = GetCursorPosition();
+	MidiNoteIterator begin(take);
+	MidiNoteIterator end = begin;
+	end.moveToEnd();
+	if (begin == end) {
+		// No notes.
+		return {-1, -1};
+	}
+	auto range = equal_range(begin, end, now);
 	// Find the first note of the chord.
-	int firstNote = direction == -1 ? count - 1 : 0;
-	int movement = direction == 0 ? 1 : direction;
-	bool found = false;
-	int nowNote = -1;
-	double firstStart;
-	for (; 0 <= firstNote && firstNote < count; firstNote += movement) {
-		MIDI_GetNote(take, firstNote, NULL, NULL, &firstStart, NULL, NULL, NULL, NULL);
-		firstStart = MIDI_GetProjTimeFromPPQPos(take, firstStart);
-		if (firstStart == now)
-			nowNote = firstNote;
-		if ((direction == 0 && firstStart == now)
-			|| (direction == 1 && firstStart > now)
-			|| (direction == -1 && firstStart < now)
-		) {
-			found = true;
-			break;
-		}
-	}
-	if (!found) {
-		if (nowNote != -1) {
-			// No chord in the requested direction, so use the chord at the cursor.
-			firstNote = nowNote;
-			firstStart = now;
-			// Reverse the direction to find the remaining notes in the chord.
-			movement = direction == 1 ? -1 : 1;
-		} else
-			return {-1, -1};
+	MidiNoteIterator firstNote = end;
+	if (direction == 1 && range.second != end) {
+		// Return chord after the cursor.
+		// range.second is the first note after now.
+		firstNote = range.second;
+	} else if (direction == -1 && range.first != begin) {
+		// Return chord before the cursor.
+		// range.first is the first note at or after now, so one before that is
+		// the first note before now.
+		firstNote = range.first;
+		--firstNote;
+	} else if (range.first != range.second) {
+		// Return chord at the cursor.
+		return {range.first.getIndex(), range.second.getIndex() - 1};
+	} else {
+		// Nothing in the requested direction or at the cursor.
+		return {-1, -1};
 	}
 	// Find the last note of the chord.
-	int lastNote = firstNote;
-	for (int note = lastNote + movement; 0 <= note && note < count; note += movement) {
-		double start;
-		MIDI_GetNote(take, note, NULL, NULL, &start, NULL, NULL, NULL, NULL);
-		start = MIDI_GetProjTimeFromPPQPos(take, start);
-		if (start != firstStart)
+	double firstStart = *firstNote;
+	MidiNoteIterator lastNote = firstNote;
+	MidiNoteIterator note = firstNote;
+	for (note += direction; begin <= note && note < end; note += direction) {
```

Comment:
I'm a bit puzzled by this. Why does this note += direction twice? Doesn't this mean note is updated twice in the loop, once at the start of the loop and one at the end?